File: Sources/Bifrost/Models/ThunderstorePackage.swift
import Foundation

/// A single mod package as returned by Thunderstore's v1 package index for
/// the Valheim community. Only the fields Bifrost actually uses are
/// decoded; everything else in the response is ignored.
struct ThunderstorePackage: Codable, Identifiable, Sendable, Hashable {
    let name: String
    let fullName: String
    let owner: String
    let packageURL: URL
    let dateUpdated: Date
    let ratingScore: Int
    let isDeprecated: Bool
    let categories: [String]
    let versions: [Version]

    var id: String { fullName }

    /// The most recent version, which the index always lists first.
    var latestVersion: Version? { versions.first }

    enum CodingKeys: String, CodingKey {
        case name
        case fullName = "full_name"
        case owner
        case packageURL = "package_url"
        case dateUpdated = "date_updated"
        case ratingScore = "rating_score"
        case isDeprecated = "is_deprecated"
        case categories
        case versions
    }

    struct Version: Codable, Sendable, Hashable {
        let name: String
        let fullName: String
        let description: String
        let icon: String?
        let versionNumber: String
        /// Dependency identifiers in "Author-Name-Version" form.
        let dependencies: [String]
        let downloadURL: URL
        let downloads: Int
        let fileSize: Int

        enum CodingKeys: String, CodingKey {
            case name
            case fullName = "full_name"
            case description
            case icon
            case versionNumber = "version_number"
            case dependencies
            case downloadURL = "download_url"
            case downloads
            case fileSize = "file_size"
        }
    }
}

extension ThunderstorePackage {
    /// Total downloads across all versions — what the Thunderstore site
    /// itself shows as the package's download count.
    var totalDownloads: Int {
        versions.reduce(0) { $0 + $1.downloads }
    }
}

File: windows/src/Bifrost.Core/Models/ThunderstorePackage.cs
using System.Text.Json.Serialization;

namespace Bifrost.Core.Models;

/// <summary>
/// A single mod package as returned by Thunderstore's v1 package index for
/// the Valheim community. Only the fields Bifrost actually uses are decoded;
/// everything else in the response is ignored. Mirrors the macOS reference
/// implementation's <c>ThunderstorePackage.swift</c>.
/// </summary>
public sealed class ThunderstorePackage
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("full_name")]
    public string FullName { get; set; } = string.Empty;

    [JsonPropertyName("owner")]
    public string Owner { get; set; } = string.Empty;

    [JsonPropertyName("package_url")]
    public string PackageUrl { get; set; } = string.Empty;

    [JsonPropertyName("date_updated")]
    public DateTimeOffset DateUpdated { get; set; }

    [JsonPropertyName("rating_score")]
    public int RatingScore { get; set; }

    [JsonPropertyName("is_deprecated")]
    public bool IsDeprecated { get; set; }

    [JsonPropertyName("categories")]
    public List<string> Categories { get; set; } = new();

    [JsonPropertyName("versions")]
    public List<Version> Versions { get; set; } = new();

    /// <summary>The most recent version, which the index always lists first.</summary>
    [JsonIgnore]
    public Version? LatestVersion => Versions.Count > 0 ? Versions[0] : null;

    /// <summary>Total downloads across all versions.</summary>
    [JsonIgnore]
    public int TotalDownloads => Versions.Sum(v => v.Downloads);

    public sealed class Version
    {
        [JsonPropertyName("name")]
        public string Name { get; set; } = string.Empty;

        [JsonPropertyName("full_name")]
        public string FullName { get; set; } = string.Empty;

        [JsonPropertyName("description")]
        public string Description { get; set; } = string.Empty;

        [JsonPropertyName("icon")]
        public string? Icon { get; set; }

        [JsonPropertyName("version_number")]
        public string VersionNumber { get; set; } = string.Empty;

        /// <summary>Dependency identifiers in "Author-Name-Version" form.</summary>
        [JsonPropertyName("dependencies")]
        public List<string> Dependencies { get; set; } = new();

        [JsonPropertyName("download_url")]
        public string DownloadUrl { get; set; } = string.Empty;

        [JsonPropertyName("downloads")]
        public int Downloads { get; set; }

        [JsonPropertyName("file_size")]
        public long FileSize { get; set; }
    }
}
